8343789: Move mutable nmethod data out of CodeCache

bulasevich opened the pull request:
This change relocates mutable data (such as relocations, metadata, jvmci data) from the nmethod. The change follows the recent PR #18984, which relocated immutable nmethod data out of the CodeCache.

OOPs was initially moved to a new mutable data blob, but then moved back to nmethod due to performance issues on dacapo benchmarks on aarch with ShenandoagGC (why Shenandoah: it is the only GC with supports_instruction_patching=false - it requires loading from the oops table in compiled code, which takes three instructions for a remote data).

Although performance is not the main focus, testing on AArch64 CPUs, where code density plays a significant role, has shown a 1–2% performance improvement in specific scenarios, such as the CodeCacheStress test and the Renaissance Dotty benchmark.

The numbers. Immutable data constitutes **~30%** on the nmehtod. Mutable data constitutes **~8%** of nmethod. Example (statistics collected on the CodeCacheStress benchmark):
- nmethod_count:134000, total_compilation_time: 510460ms
- total allocation time malloc_mutable/malloc_immutable/CodeCache_alloc: 62ms/114ms/6333ms,
- total allocation size (mutable/immutable/nmentod): 64MB/192MB/488MB

Functional testing: jtreg on arm/aarch/x86.
Performance testing: renaissance/dacapo/SPECjvm2008 benchmarks.

Alternative solution (see comments): In the future, relocations can be moved to _immutable_data.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8343789](https://bugs.openjdk.org/browse/JDK-8343789): Move mutable nmethod data out of CodeCache (**Enhancement** - P4)


### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21276/head:pull/21276` \
`$ git checkout pull/21276`

Update a local copy of the PR: \
`$ git checkout pull/21276` \
`$ git pull https://git.openjdk.org/jdk.git pull/21276/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21276`

View PR using the GUI difftool: \
`$ git pr show -t 21276`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21276.diff">https://git.openjdk.org/jdk/pull/21276.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21276#issuecomment-2491335871)
</details>

bulasevich commented:
@vnkozlov Hi Vladimir,
What do you think about the idea of ​​moving relocInfo data out of nmethod additionally to recent [Move immutable nmethod data from CodeCache](https://github.com/openjdk/jdk/pull/18984)? It would reduce the CodeHeap fill by 5%.

vnkozlov commented on src/hotspot/share/code/codeBlob.cpp:
```diff
@@ -62,32 +62,42 @@ unsigned int CodeBlob::align_code_offset(int offset) {
 CodeBlob::CodeBlob(const char* name, CodeBlobKind kind, CodeBuffer* cb, int size, uint16_t header_size,
-                   int16_t frame_complete_offset, int frame_size, OopMapSet* oop_maps, bool caller_must_gc_arguments) :
+                   int16_t frame_complete_offset, int frame_size, OopMapSet* oop_maps, bool caller_must_gc_arguments,
+                   bool external_mutable_data) :
```
I suggest to add `assert(!external_mutable_data || (kind == CodeBlobKind::Nmethod)`

vnkozlov commented on src/hotspot/share/code/codeBlob.hpp:
```diff
@@ -124,13 +124,17 @@ class CodeBlob {
 
   bool _caller_must_gc_arguments;
 
+  address _mutable_data;
+  int     _mutable_data_size;
+
```
Should we add special CodeBlob subclass for nmethod to avoid increase of size for all blobs and stubs?

bulasevich replied:
I am not sure. All CodeBlobs with relocation info needs a mutable data. Let me know if you think it must be a separate subclass.

vnkozlov commented on src/hotspot/share/runtime/vmStructs.cpp:
```diff
@@ -592,6 +592,8 @@
   nonstatic_field(nmethod,                     _osr_entry_point,                              address)                               \
   nonstatic_field(nmethod,                     _immutable_data,                               address)                               \
   nonstatic_field(nmethod,                     _immutable_data_size,                          int)                                   \
+  nonstatic_field(nmethod,                     _mutable_data,                                 address)                               \
+  nonstatic_field(nmethod,                     _mutable_data_size,                            int)                                   \
```
They are filed in CodeBlob and not in nmethod.

vnkozlov reviewed (commented):
First, thank you for doing his work.

Main question is: why you did it only for `nmethod`?

Second question: do you see any performance effects with this change?
My concern is that we iterate relocation info data from different memory space to patch code.

vnkozlov commented:
Note, with https://bugs.openjdk.org/browse/JDK-8334691 and other changes I moving into direction to make relocation info data immutable. It is already "immutable" in mainline JDK after https://bugs.openjdk.org/browse/JDK-8333819. But it is still mutable in Leyden because we have to patch indexes during publishing nmethod.

My idea was to move relocation info data (which has big size) into `immutable` data section of nmethod. And leave mutable `_oops` and `_metadata` together with code since they are relatively small and we need to patch them together with code.

vnkozlov commented:
Mutable sizes % do not add up:
```
mutable data    = 6071648 (9.396180%)
   relocation    = 3437176 (12.846409%)
   oops          = 239488 (0.895084%)
   metadata      = 2394984 (8.951227%)
```

bulasevich commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:
```diff
@@ -5270,9 +5270,8 @@ void MacroAssembler::movoop(Register dst, jobject obj) {
   } else {
     address dummy = address(uintptr_t(pc()) & -wordSize); // A nearby aligned address
-    ldr_constant(dst, Address(dummy, rspec));
+    mov(dst, Address(dummy, rspec));
```
- it is not a load from a Constant Pool, so calling ldr_constant here is seems incorrect
- the ldr_constant function utilizes either ldr (with a range limit of ±1MB) or, when -XX:-NearCpool is enabled, adrp (range limit of ±2GB) followed by ldr — both of which may fall short when mutable data is allocated on the C heap.

dean-long replied:
This change looks wrong, for a number of reasons.  First, the dummy address would no longer be needed, and we could just use the same mov as the supports_instruction_patching() case.  However, if supports_instruction_patching() is false, I think we are not allowed to generate a multi-instruction movz/movk sequence.  We really need something like ldr_constant for this case, so that we load from memory.
However, as you point out, this is tied to NearCpool.  For a far metadata slot access, ADR+LDR is the right answer.  After this change, will there be any metadata left that could still benefit from NearCpool?   If not, then it might make sense to turn it off permanently.  Instead of choosing between PC-relative "ldr literal" and far ADR+LDR based on NearCpool, we could decide based on the distance to the metadata table. I believe "ldr literal" only has a 1MB range.
CC @theRealAph

bulasevich replied:
That's right. Thanks for pointing that out to me.
I have a fix for movoop issue on supports_instruction_patching=false case. Probably it should be considered as a separate change: https://github.com/openjdk/jdk/pull/22448

bulasevich commented:
> Main question is: why you did it only for nmethod?

Yes. I did it symmetrically to a separate immutable data storage for nmethod. Now I see I do not like implementation where for some blobs relocation info is local, but for other it stays aside. I am going to rework that.

> any performance effects with this change?

For now, I do not see any performance difference with a benchmark I'm used to. I need to do more benchmarking.

bulasevich commented:
> Mutable sizes % do not add up:

Thanks. The correct sizes:
```
Statistics for 8171 bytecoded nmethods for C1:
 mutable data    = 10488856 (8.688409%)
   relocation    = 6573064 (62.667118%)
   oops          = 515680 (4.916456%)
   metadata      = 3400112 (32.416424%)
Statistics for 8171 bytecoded nmethods for C2:
 mutable data    = 6572064 (10.118859%)
   relocation    = 3406216 (51.828709%)
   oops          = 305912 (4.654733%)
   metadata      = 2859936 (43.516556%)
```

bulasevich commented:
> My idea was to move relocation info data (which has big size) into `immutable` data section of nmethod. And leave mutable `_oops` and `_metadata` together with code since they are relatively small and we need to patch them together with code.

Hmm. If relocation info goes to an immutable blob, oops+metadata hardly deserves a separate blob.

bulasevich commented:
Performance update. On an aarch machine the CodeCacheStress benchmark shows a 1-2% performance improvement with this change,

Statistics on the CodeCacheStress benchmark with high numberOfClasses-instanceCount-rangeOfClasses parameter values:
- nmethod_count:134000, total_compilation_time: 510460ms
- total allocation time malloc_mutable/malloc_immutable/CodeCache_alloc: 62ms/114ms/6333ms, 
- total allocation size (mutable/immutable/nmentod): 64MB/192MB/488MB

dean-long commented:
It would be nice to make relocations immutable, but one roadblock is the use of relocInfo::change_reloc_info_for_address() by C1 patching.  We would need to separate mutable and immutable relocations, or replace C1 patching with deoptimization, like on DEOPTIMIZE_WHEN_PATCHING aarch64.

dean-long commented on src/hotspot/share/code/codeBlob.cpp:
```diff
@@ -63,42 +63,51 @@ unsigned int CodeBlob::align_code_offset(int offset) {
   assert(_frame_size >= -1, "must use frame size or -1 for runtime stubs");
 #endif // COMPILER1
 
+  // The mutable_data_size is either calculated by the nmethod constructor to account
+  // for reloc_info and additional data, or it is set here to accommodate only the relocation data.
+  _mutable_data_size = (mutable_data_size == 0) ? cb->total_relocation_size() : mutable_data_size;
```
This seems strange to treat relocations as special.  Wouldn't it be better to have the caller always pass in the correct value?

dean-long replied:
Or compute using something like required_mutable_data_space()?

bulasevich replied:
Alright. Thank you. I moved mutable_data_size calculation out of CodeBlob.

dean-long commented on src/hotspot/share/code/codeBlob.hpp:
```diff
@@ -108,8 +108,6 @@ class CodeBlob {
   int      _relocation_size;       // size of relocation (could be bigger than 64Kb)
```
For offsets into the external mutable/immutable data, we could reduce codecache footprint further by moving these into a a header section of the external data block.  That also allows those blocks to be self-describing, which could help with error reporting or debugging.

bulasevich replied:
Sounds reasonable. But the downside is that in this case the oops iterator needs an additional load (nmethod->mutable_data->oop_size) to check if the oops list is empty.

dean-long commented on src/hotspot/share/code/codeBlob.hpp:
```diff
@@ -124,13 +122,17 @@ class CodeBlob {
   CodeBlob(const char* name, CodeBlobKind kind, CodeBuffer* cb, int size, uint16_t header_size,
-           int16_t frame_complete_offset, int frame_size, OopMapSet* oop_maps, bool caller_must_gc_arguments);
+           int16_t frame_complete_offset, int frame_size, OopMapSet* oop_maps, bool caller_must_gc_arguments,
+           int mutable_data_size = 0);
```
If we want to allow the default for mutable data size to be the relocations size, then instead of using = 0 here, you could do this instead:
```
 CodeBlob(const char* name, CodeBlobKind kind, CodeBuffer* cb, int size, uint16_t header_size,
           int16_t frame_complete_offset, int frame_size, OopMapSet* oop_maps, bool caller_must_gc_arguments,
           int mutable_data_size);

 CodeBlob(const char* name, CodeBlobKind kind, CodeBuffer* cb, int size, uint16_t header_size,
           int16_t frame_complete_offset, int frame_size, OopMapSet* oop_maps, bool caller_must_gc_arguments) : 
    CodeBlob(name, kind, cb, size, header_size,
           frame_complete_offset, frame_size, oop_maps, caller_must_gc_arguments,
           cb->total_relocation_size)
{
}
```
but I would prefer not to treat relocations as special, and have the caller always pass the correct value.

bulasevich replied:
Agree.

dean-long commented on src/hotspot/share/code/nmethod.cpp:
```diff
@@ -2127,9 +2149,13 @@ void nmethod::purge(bool unregister_nmethod) {
   delete[] _compiled_ic_data;
 
-  if (_immutable_data != data_end()) {
+  if (_immutable_data != blob_end()) {
```
Is this just a name change, or a semantic change?

bulasevich replied:
In several places _immutable_data is set to data_end() "valid not null address" address which was actually the end of the code blob. With this change I remove the _data part of the code blob as well as the data_begin() and data_end() functions. I think blob_end() is a good replacement for data_end() for empty _immutable_data cases.

dean-long commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp:
```diff
@@ -1416,11 +1416,11 @@ class MacroAssembler: public Assembler {
     } else {
       uint64_t offset;
-      adrp(dest, InternalAddress(const_addr.target()), offset);
+      adrp(dest, const_addr, offset);
```
I don't see how this ADRP path ever gets called now.  The only caller is in MacroAssembler::movoop(), which uses a dummy Address in the CodeCache.  I think we need to force near/far with an extra bool parameter.  The way this function is currently used, a better name might be ldr_patchable().

bulasevich replied:
Thanks for pointing that out. I reworked the method a little.

ldr_patchable() is only called by movoop(). If oops is moved to a separate location, the distance is certainly > 1MB and adrp+ldr is the only way to access oops. The single LDR path is not used now, but I leave it for future use.

```
  void ldr_patchable(Register dest, const Address &const_addr, bool fits_in_ldr_range = false) {
    if (fits_in_ldr_range) {
      intptr_t offset = pc() - const_addr.target();
      assert(offset >= -1024 * 1024, offset < 1024 * 1024, "pointer does not fit into pc-relative ldr range")
      ldr(dest, const_addr);
    } else {
      uint64_t offset;
      adrp(dest, const_addr, offset);
      ldr(dest, Address(dest, offset));
    }
  }
```

dean-long replied:
The problem with leaving "fits_in_ldr_range" in for future use is that it hasn't been tested.  I suggest just removing it.  It uses pc() instead of the codecache boundaries to decide the range, so it only works for code that isn't relocated.

bulasevich replied:
Agree. Thank you!

dean-long commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp:
```diff
@@ -1415,12 +1415,12 @@ class MacroAssembler: public Assembler {
-  void ldr_constant(Register dest, const Address &const_addr) {
-    if (NearCpool) {
+  void ldr_patchable(Register dest, const Address &const_addr) {
+    if (CodeCache::contains(const_addr.target())) {
```
Doesn't this cause us to generate LDR for oops outside the code cache, when what we need is the ADRP below?  The caller is still using a PC-relative dummy address.

bulasevich replied:
(written above) in ldr_patchable adrp+ldr path is used.

Code generator encodes ldr to dummy address (real address is stored in oop_Relocation). During relocation, adrp+ldr are patched to a real address. This code only works for UseShenandoahGC and ldr_patchable only uses the adrp+ldr path.

dean-long reviewed (approved):
This looks good now, but I just noticed the old code would have replaced the oop_Relocation with a internal_word_Relocation for the NearCpool==false case.  How did that ever work correctly?

dean-long commented:
/reviewers 2

bulasevich commented:
> I just noticed the old code would have replaced the oop_Relocation with a internal_word_Relocation for the NearCpool==false case. How did that ever work correctly?

Thanks for review!

NearCpool==false was not functioning correctly. I had considered a separate [fix](https://github.com/openjdk/jdk/pull/22448/files) for it, but with the current change, the incorrect code is eliminated.

```
$ ~/jdk-23/bin/java -XX:-NearCpool -XX:+UseShenandoahGC -version
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x0000ffff6f94c190, pid=95263, tid=95264
#
# JRE version: OpenJDK Runtime Environment (23.0+38) (build 23+38)
# Java VM: OpenJDK 64-Bit Server VM (23+38, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, shenandoah gc, linux-aarch64)
# Problematic frame:
# j  java.lang.module.ModuleDescriptor.<init>(Ljava/lang/String;Ljava/lang/module/ModuleDescriptor$Version;Ljava/util/Set;Ljava/util/Set;Ljava/util/Set;Ljava/util/Set;Ljava/util/Set;Ljava/util/Set;Ljava/util/Set;Ljava/lang/String;IZ)V+29 java.base
#
# Core dump will be written.
```

TobiHartmann commented:
I see this failure in our testing with an internal test:

```
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (/workspace/open/src/hotspot/share/code/nmethod.cpp:1513), pid=1863674, tid=1863692
#  assert(static_cast<int>(_metadata_offset) == reloc_size + oop_size) failed: failed: 21744 != 87280

Current CompileTask:
C2:61370 3693    b  4       org.apache.lucene.analysis.en.KStemData1::<clinit> (27623 bytes)

Stack: [0x00007f3acf3f5000,0x00007f3acf4f5000],  sp=0x00007f3acf4f0a50,  free space=1006k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x14d9151]  nmethod::nmethod(Method*, CompilerType, int, int, int, int, int, unsigned char*, CodeOffsets*, int, DebugInformationRecorder*, Dependencies*, CodeBuffer*, int, OopMapSet*, ExceptionHandlerTable*, ImplicitExceptionTable*, AbstractCompiler*, CompLevel, char*, int, JVMCINMethodData*)+0x981  (nmethod.cpp:1513)
V  [libjvm.so+0x14d94df]  nmethod::new_nmethod(methodHandle const&, int, int, CodeOffsets*, int, DebugInformationRecorder*, Dependencies*, CodeBuffer*, int, OopMapSet*, ExceptionHandlerTable*, ImplicitExceptionTable*, AbstractCompiler*, CompLevel, char*, int, JVMCINMethodData*)+0x22f  (nmethod.cpp:1200)
V  [libjvm.so+0x926cb4]  ciEnv::register_method(ciMethod*, int, CodeOffsets*, int, CodeBuffer*, int, OopMapSet*, ExceptionHandlerTable*, ImplicitExceptionTable*, AbstractCompiler*, bool, bool, bool, bool, int)+0x4c4  (ciEnv.cpp:1063)
V  [libjvm.so+0x157b0d5]  PhaseOutput::install_code(ciMethod*, int, AbstractCompiler*, bool, bool)+0x125  (output.cpp:3443)
V  [libjvm.so+0xa54d02]  Compile::Code_Gen()+0x612  (compile.cpp:3033)
V  [libjvm.so+0xa579cf]  Compile::Compile(ciEnv*, ciMethod*, int, Options, DirectiveSet*)+0x1c6f  (compile.cpp:885)
V  [libjvm.so+0x89f495]  C2Compiler::compile_method(ciEnv*, ciMethod*, int, bool, DirectiveSet*)+0x1d5  (c2compiler.cpp:142)
V  [libjvm.so+0xa63a88]  CompileBroker::invoke_compiler_on_method(CompileTask*)+0x928  (compileBroker.cpp:2319)
V  [libjvm.so+0xa647c8]  CompileBroker::compiler_thread_loop()+0x528  (compileBroker.cpp:1977)
V  [libjvm.so+0xf3478e]  JavaThread::thread_main_inner()+0xee  (javaThread.cpp:777)
V  [libjvm.so+0x1880536]  Thread::call_run()+0xb6  (thread.cpp:232)
V  [libjvm.so+0x155a188]  thread_native_entry(Thread*)+0x128  (os_linux.cpp:860)
```

TobiHartmann commented:
Looks like this also triggers with `compiler/escapeAnalysis/TestFindInstMemRecursion.java` on Linux x64 and `-XX:-TieredCompilation -XX:+StressReflectiveCode -XX:-ReduceInitialCardMarks -XX:-ReduceBulkZeroing -XX:-ReduceFieldZeroing`. This change needs more testing before integration.

bulasevich commented:
@TobiHartmann 

> I see this failure in our testing with an internal test

Thanks for catching that issue! I've addressed it.

> This change needs more testing before integration.

Thank you for pointing that out. Here is my test plan for additional validation. Please let me know if you have any suggestions or if adjustments are needed:

Functional Testing:
- Platforms: x86 and AArch64
- Build Configurations: Release, Fastdebug, and Slowdebug
- Java Options: Default and -XX:+UseShenandoahGC
- Tests: jtreg tests covering Hotspot and JDK, tiers 1–3

Performance Testing:
- Benchmarks: DaCapo and Renaissance
- Modes: Default and -XX:+UseShenandoahGC

TobiHartmann commented:
Sounds good. I'll run some testing as well and will report back once it passed.

vnkozlov commented:
@bulasevich, please wait my review before integration.

theRealAph commented:
> replace C1 patching with deoptimization, like on DEOPTIMIZE_WHEN_PATCHING aarch64.

It might be worth looking at that. My experiments at the time the AArch64 port was done indicated that only 10% of deoptimization on AArch64 was caused by patching events. Most of the 90% remaining was tiered events.

It is possible to generate C1 code for AArch64 that is patchable, but the frequent additional indirections in generated code are worse for performance than the occasional deoptimization.

TobiHartmann commented:
I see more failures in testing:

compiler/jsr292/CallSiteDepContextTest.java fails on Windows with -XX:+UseZGC:

```
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (workspace\open\src\hotspot\share\code\dependencies.cpp:2038), pid=109676, tid=68072
#  assert(call_site->is_a(vmClasses::CallSite_klass())) failed: sanity

Current thread (0x000002b4343ac600):  JavaThread "MainThread"        [_thread_in_vm, id=68072, stack(0x000000f121300000,0x000000f121400000) (1024K)]

Stack: [0x000000f121300000,0x000000f121400000],  sp=0x000000f1213fe9f0,  free space=1018k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [jvm.dll+0x601c63]  Dependencies::check_call_site_target_value+0x3e3  (dependencies.cpp:2038)
V  [jvm.dll+0x60182f]  Dependencies::DepStream::check_call_site_dependency+0x6f  (dependencies.cpp:2145)
V  [jvm.dll+0xd0898d]  nmethod::check_dependency_on+0x7d  (nmethod.cpp:2853)
V  [jvm.dll+0x609a19]  DependencyContext::mark_dependent_nmethods+0xb9  (dependencyContext.cpp:74)
V  [jvm.dll+0xcd2ff2]  MethodHandles::mark_dependent_nmethods+0xa2  (methodHandles.cpp:953)
V  [jvm.dll+0xccbf2e]  MHN_setCallSiteTargetNormal+0x1ee  (methodHandles.cpp:1211)
C  0x000002b02377ebac  (no source info available)
```

Same test also fails with:

```
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (/workspace/open/src/hotspot/share/oops/compressedKlass.inline.hpp:88), pid=95938, tid=35843
#  assert(nk >= _lowest_valid_narrow_klass_id && nk <= _highest_valid_narrow_klass_id) failed: narrowKlass ID out of range (3131947710)

---------------  T H R E A D  ---------------

Current thread (0x0000000142865e10):  JavaThread "MainThread"        [_thread_in_vm, id=35843, stack(0x0000000170c4c000,0x0000000170e4f000) (2060K)]

Stack: [0x0000000170c4c000,0x0000000170e4f000],  sp=0x0000000170e4dd00,  free space=2055k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.dylib+0x1161294]  VMError::report(outputStream*, bool)+0x1aac  (compressedKlass.inline.hpp:88)
V  [libjvm.dylib+0x1164880]  VMError::report_and_die(int, char const*, char const*, char*, Thread*, unsigned char*, void const*, void const*, char const*, int, unsigned long)+0x548
V  [libjvm.dylib+0x576b0c]  print_error_for_unit_test(char const*, char const*, char*)+0x0
V  [libjvm.dylib+0x70a898]  CompressedKlassPointers::check_encodable(void const*)+0x0
V  [libjvm.dylib+0x585d24]  oopDesc::klass() const+0x104
V  [libjvm.dylib+0x5afbec]  Dependencies::check_call_site_target_value(oop, oop, CallSiteDepChange*)+0xcc
V  [libjvm.dylib+0x5b09e4]  Dependencies::DepStream::check_call_site_dependency(CallSiteDepChange*)+0x84
V  [libjvm.dylib+0xde663c]  nmethod::check_dependency_on(DepChange&)+0x50
V  [libjvm.dylib+0x5b291c]  DependencyContext::mark_dependent_nmethods(DeoptimizationScope*, DepChange&)+0xa4
V  [libjvm.dylib+0xda4dfc]  MethodHandles::mark_dependent_nmethods(DeoptimizationScope*, Handle, Handle)+0xec
V  [libjvm.dylib+0xda7ea4]  MHN_setCallSiteTargetNormal+0x304
j  java.lang.invoke.MethodHandleNatives.setCallSiteTargetNormal(Ljava/lang/invoke/CallSite;Ljava/lang/invoke/MethodHandle;)V+0 java.base@25-internal
j  java.lang.invoke.CallSite.setTargetNormal(Ljava/lang/invoke/MethodHandle;)V+7 java.base@25-internal
j  java.lang.invoke.MutableCallSite.setTarget(Ljava/lang/invoke/MethodHandle;)V+2 java.base@25-internal
j  compiler.jsr292.CallSiteDepContextTest.testGC(ZZ)V+267
j  compiler.jsr292.CallSiteDepContextTest.main([Ljava/lang/String;)V+11
j  java.lang.invoke.LambdaForm$DMH+0x00000c0000001c00.invokeStatic(Ljava/lang/Object;Ljava/lang/Object;)V+10 java.base@25-internal
j  java.lang.invoke.LambdaForm$MH+0x00000c0000003400.invoke(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;+33 java.base@25-internal
j  java.lang.invoke.Invokers$Holder.invokeExact_MT(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;+20 java.base@25-internal
j  jdk.internal.reflect.DirectMethodHandleAccessor.invokeImpl(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;+55 java.base@25-internal
j  jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;+23 java.base@25-internal
j  java.lang.reflect.Method.invoke(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;+102 java.base@25-internal
j  com.sun.javatest.regtest.agent.MainWrapper$MainTask.run()V+134
j  java.lang.Thread.runWith(Ljava/lang/Object;Ljava/lang/Runnable;)V+5 java.base@25-internal
j  java.lang.Thread.run()V+19 java.base@25-internal
v  ~StubRoutines::call_stub 0x00000001160d0180
V  [libjvm.dylib+0x88beec]  JavaCalls::call_helper(JavaValue*, methodHandle const&, JavaCallArguments*, JavaThread*)+0x448
V  [libjvm.dylib+0x88ac08]  JavaCalls::call_virtual(JavaValue*, Klass*, Symbol*, Symbol*, JavaCallArguments*, JavaThread*)+0x1c4
V  [libjvm.dylib+0x88ada4]  JavaCalls::call_virtual(JavaValue*, Handle, Klass*, Symbol*, Symbol*, JavaThread*)+0x6c
V  [libjvm.dylib+0xa1289c]  thread_entry(JavaThread*, JavaThread*)+0x12c
V  [libjvm.dylib+0x8c2088]  JavaThread::thread_main_inner()+0x1a8
V  [libjvm.dylib+0x10ae844]  Thread::call_run()+0xf4
V  [libjvm.dylib+0xe467f0]  thread_native_entry(Thread*)+0x138
C  [libsystem_pthread.dylib+0x6f94]  _pthread_start+0x88
Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
j  java.lang.invoke.MethodHandleNatives.setCallSiteTargetNormal(Ljava/lang/invoke/CallSite;Ljava/lang/invoke/MethodHandle;)V+0 java.base@25-internal
j  java.lang.invoke.CallSite.setTargetNormal(Ljava/lang/invoke/MethodHandle;)V+7 java.base@25-internal
j  java.lang.invoke.MutableCallSite.setTarget(Ljava/lang/invoke/MethodHandle;)V+2 java.base@25-internal
j  compiler.jsr292.CallSiteDepContextTest.testGC(ZZ)V+267
j  compiler.jsr292.CallSiteDepContextTest.main([Ljava/lang/String;)V+11
j  java.lang.invoke.LambdaForm$DMH+0x00000c0000001c00.invokeStatic(Ljava/lang/Object;Ljava/lang/Object;)V+10 java.base@25-internal
j  java.lang.invoke.LambdaForm$MH+0x00000c0000003400.invoke(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;+33 java.base@25-internal
j  java.lang.invoke.Invokers$Holder.invokeExact_MT(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;+20 java.base@25-internal
j  jdk.internal.reflect.DirectMethodHandleAccessor.invokeImpl(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;+55 java.base@25-internal
j  jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;+23 java.base@25-internal
j  java.lang.reflect.Method.invoke(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;+102 java.base@25-internal
j  com.sun.javatest.regtest.agent.MainWrapper$MainTask.run()V+134
j  java.lang.Thread.runWith(Ljava/lang/Object;Ljava/lang/Runnable;)V+5 java.base@25-internal
j  java.lang.Thread.run()V+19 java.base@25-internal
v  ~StubRoutines::call_stub 0x00000001160d0180
Lock stack of current Java thread (top to bottom):
```

TobiHartmann commented:
@stefank just pointed out that this could be a regression from [JDK-8347564](https://bugs.openjdk.org/browse/JDK-8347564) that just went in an unrelated to this change.

theRealAph commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp:
```diff
@@ -1415,14 +1415,11 @@ class MacroAssembler: public Assembler {
+    uint64_t offset;
+    bool force_movk = true; // movk is important if the target can be more than 4GB away
+    adrp(dest, const_addr, offset, force_movk);
+    ldr(dest, Address(dest, offset));
```
I wonder if this really is the best way to do it. It's not clear to me that there is any advantage of using `adrp` in this case rather than a simple `mov(scratch, const_adr); ldr(dest, Address(scratch);`. The `mov` would produce `movz; movk; movk` which almost certainly execute in a single cycle, then a load without an offset, which is a single micro-op rather than two micro-ops for load+offset. All we've gained for this complication is a small reduction in code density rather than a performance improvement. I'd go with simplicity.

adinn replied:
Yes, I agree.

bulasevich replied:
> It's not clear to me that there is any advantage of using adrp

I think ADP+MOVK is better both in terms of performance and code density.

<details>
<summary>Simple asm experiment shows that ADRP+MOVK performs better on my machine</summary>

    $ gcc test.cpp && ./a.out

    Allocated at: 0x10ffff0000
    Elapsed time (movz-movk-movk): 3145591
    Elapsed time (adrp-movk): 2739354

    ============================================================
    #include <stdio.h>
    #include <stdlib.h>
    #include <sys/mman.h>
    #include <unistd.h>
    #include <time.h>

    int main() {
        void *desired_addr = (void *)0x10ffff0000;
        size_t size = 4096;

        int32_t* data = (int32_t*)mmap(desired_addr, size, PROT_READ | PROT_WRITE,
                                       MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED_NOREPLACE, -1, 0);
        if (data == MAP_FAILED) {
            perror("mmap failed");
            return 1;
        }
        data[0] = 13; data[1] = 14; data[2] = 15; data[3] = 16;

        printf("Allocated at: %p\n", data);
        int32_t* ptr = (int32_t*)&main;

        int aa = 1;
        int bb = 2;
        int cc = 3;
        int dd = 4;
        clock_t start;

        start = clock();
        for (int i=0; i<1000*1000*1000; i++) {
          asm (
          "movz %0, 0x10, lsl #32; movk %0, 0xffff, lsl #16; movk %0, 0; ldr %0, [%0]; "
          "movz %1, 0x10, lsl #32; movk %1, 0xffff, lsl #16; movk %1, 4; ldr %1, [%1]; "
          "movz %2, 0x10, lsl #32; movk %2, 0xffff, lsl #16; movk %2, 8; ldr %2, [%2]; "
          "movz %3, 0x10, lsl #32; movk %3, 0xffff, lsl #16; movk %3, 12;ldr %3, [%3]; "
          : "=r" (aa), "=r" (bb), "=r" (cc), "=r" (dd)  /* Output operands */
          :
          : "cc");
        }
        printf("Elapsed time (movz-movk-movk): %li\n", (clock() - start));

        start = clock();
        for (int i=0; i<1000*1000*1000; i++) {
          asm (
          "adrp %0, main; movk %0, 0, lsl #32; ldr %0, [%0, 0x0];"
          "adrp %1, main; movk %1, 0, lsl #32; ldr %1, [%1, 0x4];"
          "adrp %2, main; movk %2, 0, lsl #32; ldr %2, [%2, 0x8];"
          "adrp %3, main; movk %3, 0, lsl #32; ldr %3, [%3, 0xc];"
          : "=r" (aa), "=r" (bb), "=r" (cc), "=r" (dd)  /* Output operands */
          :
          : "cc");
        }
        printf("Elapsed time (adrp-movk): %li\n", (clock() - start));

        munmap(data, size);
        return 0;
    }

</details>

The results are consistent with llvm-mca analysis:
<details>
<summary>llvm-mca analysis suggests that ADRP+MOVK performs better on Neoverse-N1</summary>
    Neoverse-N1. ADRP-MOVK wins over MOVZ-MOVK-MOVK

    - Fewer instructions (300 vs 400)
    - Fewer total cycles (75 vs 109) - Faster execution
    - Lower instruction block throughput (0.7 vs 1.0) - More efficient execution
    - Less resource pressure - Less risk of pipeline stalls

    ===================================================================================================

    $ ../clang+llvm-18.1.0-aarch64-linux-gnu/bin/llvm-mca -mcpu=neoverse-n1 asm1.S

    Iterations:        100
    Instructions:      300
    Total Cycles:      75
    Total uOps:        300

    Dispatch Width:    8
    uOps Per Cycle:    4.00
    IPC:               4.00
    Block RThroughput: 0.7


    Instruction Info:
    [1]: #uOps
    [2]: Latency
    [3]: RThroughput
    [4]: MayLoad
    [5]: MayStore
    [6]: HasSideEffects (U)

    [1]    [2]    [3]    [4]    [5]    [6]    Instructions:
     1      1     0.33                        adrp  x0, target
     1      1     0.33                        movk  x0, #39612, lsl #32
     1      4     0.50    *                   ldr   x1, [x0]


    Resources:
    [0]   - N1UnitB
    [1.0] - N1UnitD
    [1.1] - N1UnitD
    [2.0] - N1UnitL
    [2.1] - N1UnitL
    [3]   - N1UnitM
    [4.0] - N1UnitS
    [4.1] - N1UnitS
    [5]   - N1UnitV0
    [6]   - N1UnitV1


    Resource pressure per iteration:
    [0]    [1.0]  [1.1]  [2.0]  [2.1]  [3]    [4.0]  [4.1]  [5]    [6]
     -      -      -     0.50   0.50   0.66   0.67   0.67    -      -

    Resource pressure by instruction:
    [0]    [1.0]  [1.1]  [2.0]  [2.1]  [3]    [4.0]  [4.1]  [5]    [6]    Instructions:
     -      -      -      -      -     0.33   0.33   0.34    -      -     adrp      x0, target
     -      -      -      -      -     0.33   0.34   0.33    -      -     movk      x0, #39612, lsl #32
     -      -      -     0.50   0.50    -      -      -      -      -     ldr       x1, [x0]

    ===================================================================================================

    $ ../clang+llvm-18.1.0-aarch64-linux-gnu/bin/llvm-mca -mcpu=neoverse-n1 asm2.S

    Iterations:        100
    Instructions:      400
    Total Cycles:      109
    Total uOps:        400

    Dispatch Width:    8
    uOps Per Cycle:    3.67
    IPC:               3.67
    Block RThroughput: 1.0


    Instruction Info:
    [1]: #uOps
    [2]: Latency
    [3]: RThroughput
    [4]: MayLoad
    [5]: MayStore
    [6]: HasSideEffects (U)

    [1]    [2]    [3]    [4]    [5]    [6]    Instructions:
     1      1     0.33                        mov   x0, #20014547599360
     1      1     0.33                        movk  x0, #22136, lsl #16
     1      1     0.33                        movk  x0, #39612
     1      4     0.50    *                   ldr   x1, [x0]


    Resources:
    [0]   - N1UnitB
    [1.0] - N1UnitD
    [1.1] - N1UnitD
    [2.0] - N1UnitL
    [2.1] - N1UnitL
    [3]   - N1UnitM
    [4.0] - N1UnitS
    [4.1] - N1UnitS
    [5]   - N1UnitV0
    [6]   - N1UnitV1


    Resource pressure per iteration:
    [0]    [1.0]  [1.1]  [2.0]  [2.1]  [3]    [4.0]  [4.1]  [5]    [6]
     -      -      -     0.50   0.50   0.99   1.00   1.01    -      -

    Resource pressure by instruction:
    [0]    [1.0]  [1.1]  [2.0]  [2.1]  [3]    [4.0]  [4.1]  [5]    [6]    Instructions:
     -      -      -      -      -      -     0.66   0.34    -      -     mov       x0, #20014547599360
     -      -      -      -      -     0.33   0.34   0.33    -      -     movk      x0, #22136, lsl #16
     -      -      -      -      -     0.66    -     0.34    -      -     movk      x0, #39612
     -      -      -     0.50   0.50    -      -      -      -      -     ldr       x1, [x0]

</details>



<details>
<summary>llvm-mca analysis suggests that ADRP+MOVK performs better on Neoverse-V2</summary>

    - Fewer instructions (300 vs 400)
    - Fewer total cycles (42 vs 59) - Faster execution
    - Lower instruction block throughput (0.3 vs 0.5) - More efficient execution
    - Less resource pressure - Less risk of pipeline stalls

    ===================================================================================================================================================

    $ ../clang+llvm-18.1.0-aarch64-linux-gnu/bin/llvm-mca -mcpu=neoverse-v2 asm1.S

    Iterations:        100
    Instructions:      300
    Total Cycles:      42
    Total uOps:        300

    Dispatch Width:    16
    uOps Per Cycle:    7.14
    IPC:               7.14
    Block RThroughput: 0.3


    Instruction Info:
    [1]: #uOps
    [2]: Latency
    [3]: RThroughput
    [4]: MayLoad
    [5]: MayStore
    [6]: HasSideEffects (U)

    [1]    [2]    [3]    [4]    [5]    [6]    Instructions:
     1      1     0.25                        adrp  x0, target
     1      1     0.17                        movk  x0, #39612, lsl #32
     1      4     0.33    *                   ldr   x1, [x0]


    Resources:
    [0.0] - V2UnitB
    [0.1] - V2UnitB
    [1.0] - V2UnitD
    [1.1] - V2UnitD
    [2]   - V2UnitL2
    [3.0] - V2UnitL01
    [3.1] - V2UnitL01
    [4]   - V2UnitM0
    [5]   - V2UnitM1
    [6]   - V2UnitS0
    [7]   - V2UnitS1
    [8]   - V2UnitS2
    [9]   - V2UnitS3
    [10]  - V2UnitV0
    [11]  - V2UnitV1
    [12]  - V2UnitV2
    [13]  - V2UnitV3


    Resource pressure per iteration:
    [0.0]  [0.1]  [1.0]  [1.1]  [2]    [3.0]  [3.1]  [4]    [5]    [6]    [7]    [8]    [9]    [10]   [11]   [12]   [13]
     -      -      -      -     0.33   0.33   0.34   0.33   0.33   0.34   0.34   0.33   0.33    -      -      -      -

    Resource pressure by instruction:
    [0.0]  [0.1]  [1.0]  [1.1]  [2]    [3.0]  [3.1]  [4]    [5]    [6]    [7]    [8]    [9]    [10]   [11]   [12]   [13]   Instructions:
     -      -      -      -      -      -      -     0.33   0.33   0.17   0.17    -      -      -      -      -      -     adrp     x0, target
     -      -      -      -      -      -      -      -      -     0.17   0.17   0.33   0.33    -      -      -      -     movk     x0, #39612, lsl #32
     -      -      -      -     0.33   0.33   0.34    -      -      -      -      -      -      -      -      -      -     ldr      x1, [x0]

    ===================================================================================================================================================

    $ ../clang+llvm-18.1.0-aarch64-linux-gnu/bin/llvm-mca -mcpu=neoverse-v2 asm2.S

    Iterations:        100
    Instructions:      400
    Total Cycles:      59
    Total uOps:        400

    Dispatch Width:    16
    uOps Per Cycle:    6.78
    IPC:               6.78
    Block RThroughput: 0.5


    Instruction Info:
    [1]: #uOps
    [2]: Latency
    [3]: RThroughput
    [4]: MayLoad
    [5]: MayStore
    [6]: HasSideEffects (U)

    [1]    [2]    [3]    [4]    [5]    [6]    Instructions:
     1      1     0.17                        mov   x0, #20014547599360
     1      1     0.17                        movk  x0, #22136, lsl #16
     1      1     0.17                        movk  x0, #39612
     1      4     0.33    *                   ldr   x1, [x0]


    Resources:
    [0.0] - V2UnitB
    [0.1] - V2UnitB
    [1.0] - V2UnitD
    [1.1] - V2UnitD
    [2]   - V2UnitL2
    [3.0] - V2UnitL01
    [3.1] - V2UnitL01
    [4]   - V2UnitM0
    [5]   - V2UnitM1
    [6]   - V2UnitS0
    [7]   - V2UnitS1
    [8]   - V2UnitS2
    [9]   - V2UnitS3
    [10]  - V2UnitV0
    [11]  - V2UnitV1
    [12]  - V2UnitV2
    [13]  - V2UnitV3


    Resource pressure per iteration:
    [0.0]  [0.1]  [1.0]  [1.1]  [2]    [3.0]  [3.1]  [4]    [5]    [6]    [7]    [8]    [9]    [10]   [11]   [12]   [13]
     -      -      -      -     0.33   0.33   0.34   0.50   0.50   0.50   0.50   0.50   0.50    -      -      -      -

    Resource pressure by instruction:
    [0.0]  [0.1]  [1.0]  [1.1]  [2]    [3.0]  [3.1]  [4]    [5]    [6]    [7]    [8]    [9]    [10]   [11]   [12]   [13]   Instructions:
     -      -      -      -      -      -      -      -      -     0.33   0.33   0.17   0.17    -      -      -      -     mov      x0, #20014547599360
     -      -      -      -      -      -      -     0.17   0.17   0.16   0.16   0.17   0.17    -      -      -      -     movk     x0, #22136, lsl #16
     -      -      -      -      -      -      -     0.33   0.33   0.01   0.01   0.16   0.16    -      -      -      -     movk     x0, #39612
     -      -      -      -     0.33   0.33   0.34    -      -      -      -      -      -      -      -      -      -     ldr      x1, [x0]
     
    ===================================================================================================================================================

</details>

stefank commented:
> @stefank just pointed out that this could be a regression from [JDK-8347564](https://bugs.openjdk.org/browse/JDK-8347564) that just went in.

I thought that it could have been a regression but after looking closer I don't think it is. However, I do see a problem that the code removes the GC barriers when loading oops. I'll add an inline comment.

stefank commented on src/hotspot/share/code/nmethod.cpp:
```diff
@@ -2143,14 +2159,14 @@ oop nmethod::oop_at(int index) const {
     return nullptr;
   }
-  return NMethodAccess<AS_NO_KEEPALIVE>::oop_load(oop_addr_at(index));
+  return RawAccess<>::oop_load(oop_addr_at(index));
```
This change is removing the GC barriers and is likely the cause of the ZGC crash that Tobias listed.

However, the fix is not as simple as to just reinstate the NMethodAccess call. The ZGC code uses the `oop*` to find the associated `nmethod` in the code cache. We need another way to fetch the nmethod now. So, I'm experimenting with a small change to switch out the Access API call to a direct GC barrier set call and then I pass down the `this` pointer from this function. With that you should be able skip this change.

With that said, what was the motivation for changing this?

bulasevich replied:
Right, I remember I encountered a "did not find nmethod" assertion in ZNMethod::load_oop. With my change the method fails to find nmethod by oop* which is supposed to be within the boundaries of nmethod. Based on naming semantics, I thought that if the oop wasn't in the nmethod, switching from NMethodAccess to RawAccess was acceptable. In reality, this bypasses the essential GC barrier.

Given that the nmethod is known in a call stack above, I think we should just pass it down instead of attempting a lookup.

```cpp
call stack:
- ZNMethod::load_oop(oop*, unsigned long)
- AccessInternal::PostRuntimeDispatch<ZBarrierSet::AccessBarrier<., .>, ., .>::oop_access_barrier(void*)  
- AccessInternal::RuntimeDispatch<1122372ul, oop, (AccessInternal::BarrierType)2>::load_init(void*)  
- nmethod::oop_at_phantom(int) 

oop ZNMethod::load_oop(oop* p, DecoratorSet decorators) {
  nmethod* const nm = CodeCache::find_nmethod((void*)p);
  assert(nm != nullptr, "did not find nmethod");
```

stefank replied:
Right, I created a patch to do exactly this and based it upon your PR branch. Take a look at my commit in:
https://github.com/stefank/jdk/tree/rewire_nmethod_oop_loads

I can try to get this upstreamed after I've done enough testing.

theRealAph commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:
```diff
@@ -323,6 +323,17 @@ class Patcher : public RelocActions {
     Instruction_aarch64::patch(insn_addr, 30, 29, offset_lo);
+
+    // Maybe we have a third instruction: adrp + movk + ldr with offset (e.g. ldr_patchable)
+    uint32_t insn3 = insn_at(insn_addr, 2);
```
As discussed below, it's better not to do this. `adrp()` was never intended to be used in this way, and it's not so efficient to use adrp + movk + ldr with offset rather than movz; movk; movk; ldr (no offset).

theRealAph commented:
When using Shenadoah I am still seeing embedded compressed OOPs in the instruction stream. Is that correct? These form two instructions, like so:

```
  movz(dst, 0xDEAD, 16);
  movk(dst, 0xBEEF);

```

Do you want compressed OOPs to be moved out of CodeCache as well as uncompressed OOPs? If so, you should change `loadConNNode`in C2.

vnkozlov commented on src/hotspot/share/code/nmethod.cpp:
```diff
@@ -3116,6 +3136,10 @@ void nmethod::print(outputStream* st) const {
                                              p2i(scopes_data_begin()),
                                              p2i(scopes_data_end()),
                                              scopes_data_size());
+  if (mutable_data_size() > 0)  st->print_cr(" mutable data   [" INTPTR_FORMAT "," INTPTR_FORMAT "] = %d",
+                                             p2i(mutable_data_begin()),
+                                             p2i(mutable_data_end()),
+                                             mutable_data_size());
```
Duplicated output. Remove.

vnkozlov commented on src/hotspot/share/code/nmethod.hpp:
```diff
@@ -135,8 +135,6 @@ class PcDescContainer : public CHeapObj<mtCode> {
 //
 // An nmethod contains:
 //  - header                 (the nmethod structure)
-//  [Relocation]
-//  - relocation information
```
Whole this comment have to be re-written to reflect new layout and additional data sections outside CodeCache.

vnkozlov reviewed (commented):
Good work. Here are my comments. And I did not look on AArch64 changes.

bulasevich commented:
> Do you want compressed OOPs to be moved out of CodeCache as well as uncompressed OOPs? If so, you should change `loadConNNode`in C2.

I have moved the OOPs table out of the CodeCache, but its contents remain unchanged - it still holds compressed or uncompressed pointers in the CodeHeap. As I understand it, I only need to adjust how OOPs are accessed without modifying anything else.  With ShenandoahGC, I pass jtreg tests with UseCompressedOops both enabled and disabled, so there doesn’t seem to be any issue. Please let me know if I’m mistaken.

bulasevich commented:
> If we do decide to do this, please give the forced movk version of adrp() a new name, and have adrp() call it.

Right. Thanks for pointing that out. For now, I'll add an adrp_movk() variant to handle the extra movk.

bulasevich commented:
> Good work. Here are my comments.

I have addressed your comments.
Thank you very much!!

vnkozlov reviewed (commented):
Looks good. I will submit testing.

bulasevich commented:
> Looks good. I will submit testing.

Thank you!

The change is not yet ready for final testing. I still need to remove my raw access workaround in nmethod::oop_at and rebase onto #23512 once it has been integrated.

vnkozlov commented:
Okay, then I will go ahead with my PR [23607](https://github.com/openjdk/jdk/pull/23607) which touches `CodeBlob` .

vnkozlov commented:
@bulasevich  my an other PR #23533 is ready. It will conflict with your changes. Are you okay if I push it first?

bulasevich commented:
> @bulasevich my an other PR #23533 is ready. It will conflict with your changes. Are you okay if I push it first?

Absolutely! Thanks!

theRealAph commented:
> > Do you want compressed OOPs to be moved out of CodeCache as well as uncompressed OOPs? If so, you should change `loadConNNode`in C2.
> 
> I have moved the OOPs table out of the CodeCache, but its contents remain unchanged - it still holds compressed or uncompressed pointers in the CodeHeap. As I understand it, I only need to adjust how OOPs are accessed without modifying anything else. With ShenandoahGC, I pass jtreg tests with UseCompressedOops both enabled and disabled, so there doesn’t seem to be any issue. Please let me know if I’m mistaken.

No, you're right. I misunderstood the scope of this PR.

bulasevich commented:
@vnkozlov, @dean-long, @theRealAph, @stefank
I have to note that my change causes a performance regression in the DaCapo luindex benchmark on AArch64 when using UseShenandoahGC mode. The regression is solely due to the additional adrp+movk instructions required to access oops from the compiled code. My intention was to reduce the nmethod size by moving oops out of the code cache. While this does reduce the space occupied by oops data, on AArch64 with Shenandoah, the extra instructions needed to access oops add back approximately 1% of the total nmethod size, effectively canceling the reduction achieved by moving oops. Given this, I now think it would be better to keep oops in nmethod while still moving other data (relocations, metadata, jvmci_data) out.

dean-long commented:
If we put the "oop pool" nearby, then we could continue to use the shorter instruction sequence.  We could put the oops in a "DataBlob" in the codecache.  With a segment codecache, they could even go in their own segment.
I suspect that many nmethods are referencing the same oops.  We could consider storing the oops in a deduplicated structure that GC could quickly scan, similar to OopStorage but reference-counted.

dean-long commented:
Also, it seems like there are two kinds of code density we should be concerned about:

1. not poluting icache lines with data
2. maximizing near calls in the codecache

For 1), aligning embedded data on cache line boundaries would help, but for 2) we probably would want to put any nearby DataBlobs in their own codecache segment.

vnkozlov commented:
@dean-long this is good idea but for separate RFE . @bulasevich please file one.

For now I agree with moving oops data back to nmethod blob.  Metaspace (klass*, method*) and relocation data are more stable: mostly updated during nmethod publishing.

bulasevich commented:
> If we put the "oop pool" nearby, then we could continue to use the shorter instruction sequence. We could put the oops in a "DataBlob" in the codecache.

That’s an interesting approach! By putting oops together, we can help the GC. However, the maximum offset for PC-relative LDR instructions in AArch64 is ±1MB, which is quite short for accessing a common DataBlob.

dean-long commented:
The ±1MB range is unfortunately small.  We would have to place DataBlob "islands" 1MB apart, which would fragment the codecache if they were shared and had a longer lifetime than the nmethod.  We could consider having a dedicated register that points to an external oop pool, but there's no guarantee that all the oops would fit in the ldr reg+off range.  It's an interesting problem.

dean-long commented:
Wouldn't most adrp+movk instructions for oops being computing the same or nearby base addresses?  We could set up a dedicated base pointer to the external oop table at the beginning of the code, then use ldr $oop_table + offset for each oop reference.  Or instead of a reserving a dedicated register that can't be used for anything else, we could allocate a regular spillable register, at the cost of worse performance if it needed to be spilled.

bulasevich commented:
As agreed, I moved oops back to nmethod, significantly reducing the change. All AArch-specific modifications (encoding of long load with adrp+movk+ldr and its relocation patching) were reverted.

Testing results:
- Builds: AArch & x86, client build, GraalVM build
- jtreg (hotspot & jdk tier1-3): G1/ZGC/Shenandoah/Xcomp/TieredStopAtLevel=3/-TieredCompilation - No regressions
- DaCapo & Renaissance benchmarks - No regressions

bulasevich commented:
Here is the PrintNMethodStatistics printout. It shows that in the application (a large Renaissance Dotty benchmark), we observe a significant reduction in CodeCache usage.

```diff
Statistics for 20625 bytecoded nmethods for C1:
 total size      = 121587728 (100%)
 in CodeCache    = 80406760 (66.130653%)
   header        = 4950000 (6.156199%)
   constants     = 640 (0.000796%)
   main code     = 69890600 (86.921303%)
   stub code     = 4923768 (6.123575%)
   oops          = 476752 (0.592925%)
 mutable data    = 10163920 (8.359330%)
   relocation    = 6810824 (67.009811%)
   metadata      = 3353096 (32.990185%)
 immutable data  = 31017048 (25.510014%)
   dependencies  = 606216 (1.954461%)
   nul chk table = 724344 (2.335309%)
   handler table = 222464 (0.717231%)
   scopes pcs    = 15817888 (50.997398%)
   scopes data   = 13646136 (43.995602%)
Statistics for 8290 bytecoded nmethods for C2 | Statistics for 8442 bytecoded nmethods for JVMCI
 total size      = 66679688 (100%)            |  total size      = 46208136 (100%)
 in CodeCache    = 26004920 (38.999763%)      |  in CodeCache    = 19489616 (42.177887%)
   header        = 1989600 (7.650860%)        |    header        = 2026080 (10.395690%)
   constants     = 1920 (0.007383%)           |    constants     = 540288 (2.772184%)
   main code     = 20949456 (80.559586%)      |    main code     = 14737620 (75.617805%)
   stub code     = 2702064 (10.390588%)       |    stub code     = 1904548 (9.772117%)
   oops          = 295560 (1.136554%)         |    oops          = 213544 (1.095681%)
 mutable data    = 6564928 (9.845469%)        |  mutable data    = 4168848 (9.021892%)
   relocation    = 3542736 (53.964584%)       |    relocation    = 1671384 (40.092228%)
                                              >    JVMCI data    = 202608 (4.860048%)
   metadata      = 3022192 (46.035416%)       |    metadata      = 2294856 (55.047726%)
 immutable data  = 34109840 (51.154766%)      |  immutable data  = 22549672 (48.800220%)
   dependencies  = 988000 (2.896525%)         |    dependencies  = 460104 (2.040402%)
   nul chk table = 554680 (1.626158%)         |    nul chk table = 618888 (2.744554%)
   handler table = 1787424 (5.240201%)        |    handler table = 20664 (0.091638%)
   scopes pcs    = 16152224 (47.353561%)      |    scopes pcs    = 10965040 (48.626163%)
   scopes data   = 14627512 (42.883556%)      |    scopes data   = 7746888 (34.354771%)
                                              >    speculations  = 2738088 (12.142474%)
```

By moving mutable data out of the CodeCache, we reduce CodeCache usage by the following percentages:
- C1: 10163920/(10163920+80406760) = 11%
- C2: 6564928/(6564928+26004920) = 20%
- JMVTI: 4168848/(4168848+19489616) = 18%

vnkozlov reviewed (commented):
This looks much better.

Please swap `matadata` and `jvmci data` in outputs according to data layout (metadata before jvmci data).   It is in `print_nmethod_stats()` and `print_on_impl()`.

Also please merge latest JDK which have SA cleanup related to compilers: https://github.com/openjdk/jdk/pull/23782

bulasevich commented:
> Please swap `matadata` and `jvmci data` in outputs ...
> 
> Also please merge latest JDK which have SA cleanup related to compilers: #23782

Yes. Thanks!

vnkozlov commented:
@bulasevich is it ready for testing now?

bulasevich commented:
> @bulasevich is it ready for testing now?

@vnkozlov  yes, it's ready for testing. Thanks!

vnkozlov reviewed (approved):
My testing tier1-7, stress, comp passed with one new failure [JDK-8351457](https://bugs.openjdk.org/browse/JDK-8351457) which is not related I think.

bulasevich commented:
Hi @dean-long,

Would you mind doing a re-review of this PR? I have reverted the movement of oops into a separate buffer, as it caused issues on AArch. All platform-specific details are now removed, making the change much simpler.

dean-long reviewed (approved):
Still looks good.

bulasevich commented:
/integrate

dougxc commented on src/hotspot/share/code/nmethod.cpp:
```diff
@@ -1486,26 +1499,24 @@ nmethod::nmethod(
-           (data_offset() + data_end_offset), nmethod_size);
+    CHECKED_CAST(_oops_size, uint16_t, align_up(code_buffer->total_oop_size(), oopSize));
+    uint16_t metadata_size = (uint16_t)align_up(code_buffer->total_metadata_size(), wordSize);
+    JVMCI_ONLY(CHECKED_CAST(_jvmci_data_size, uint16_t, align_up(compiler->is_jvmci() ? jvmci_data->size() : 0, oopSize)));
```
This cast is lossy in that `jvmci_data->size()` returns an int. It caused a `double free or corruption (out)` crash in Graal in the case where a `JVMCINMethodData` had a very long name. We've fixed this by [limiting the length of the name](https://github.com/openjdk/jdk/pull/24753) but I'm wondering if there was some special reason for this cast?

dean-long replied:
The cast was added by 8331087, which reduced the supported JVMCI data size to uint16_t.  I don't remember this issue with long names coming up during that review, so I guess we all missed it.  @dougxc please file a bug so we can track this.  It seems like JVMCINMethodData::copy should do something like truncate long names rather than blindly assuming it has enough space.  If uint16_t is unreasonably small for JVMCI nmethod data we could revert that change and make it 32 bits again.

dean-long replied:
Yes, my mistake.  I was thinking `_jvmci_data_offset` was used to compute `jvmci_data_end()`, not `jvmci_data_begin()`.

vnkozlov replied:
Yes we should use 32 bits. Even if we revert back to using _jvmci_data_offset we can use uint16_t because size of relocation (after which JVMCI data is placed) data is bigger.

bulasevich replied:
Thanks for the report. Yes, cast to uint16 is wrong. I am going to fix the issue here: https://github.com/openjdk/jdk/pull/24965
